src/structures.ts

```diff
@@ -0,0 +1,221 @@
+   * 
+   * @param piece - 
+   * @returns 
+   */
+  private separateWhitespace(piece: any):any {
+    if (piece.length > 1) {
+      const [before, firstItem] =
+        typeof piece[0] === 'string' ? piece[0].match(/^(\s*)([\s\S]+)$/)!.slice(1) : [[], piece[0]];
+      const [lastItem, after] =
+        typeof piece[piece.length - 1] === 'string'
+          ? piece[piece.length - 1].match(/^([\s\S]+?)(\s*)$/).slice(1)
+          : [piece[piece.length - 1], []];
+      const mid = piece.slice(1, piece.length - 1);
+      return [before, [firstItem, ...mid, lastItem], after];
+    }
+    
+    // single string
+    if (piece.length === 1 && typeof piece[0] === 'string' && /^\s+$/.test(piece[0])) {
+      return [piece[0]];
+    }
+  
+    if (piece.length === 1) {
+      const [before, text, after] =
+        typeof piece[0] === 'string'
+          ? piece[0].match(/^(\s*)([\s\S]+?)(\s*)$/)!.slice(1)
+          : [[], piece[0], []];
+      return [before, [text], after];
+    }
+  
+    return [];
+  }
+}
+
+/**
+ * ParsedObj contains original node and an array of ideas
+ * 
+ * @remarks
+ * Every item is an array of strings and HTML elements,
+ * a full-whitespace string or another ParsedObj.
+ * 
+ * @example Example ParsedObj contents
+ * ```ts
+ * {
+ *  node: <p class="chunk">,
+ *  ideas: [
+ *    ["Integer nec odio."],
+ *    " ",
+ *    ["Praesent ", <strong>, ", nibh elementum imperdiet."],
+ *    " ",
+ *    ["Sed cursus ante dapibus diam."],
+ *    " ",
+ *    [ParsedObj],
+ *    " ",
+ *    …
+ *  ]
+ * }
+ * ```
+ */
+export class ParsedObj {
+  node: Node;
+  ideas: IdeasItem[];
+  delimiter: Delimiter;
+
+  /**
+   * Constructor todo
+   * @param node - Original Node
+   * @param ideas - Every item is an array of strings and HTML elements, 
+   * a full-whitespace string or another ParsedObj.
+   * @param delimiter - Delimiter
+   */
+  constructor(node: Node, ideas: IdeasItem[], delimiter: Delimiter) {
+    const ideaProblems = this.listProblemParsedObjIdeas(ideas);
+  
+    if (ideaProblems.length > 0)
+      throw new Error(
+        `Invalid ideas at node ${JSON.stringify(node)}, problems: ${JSON.stringify(ideaProblems)}.`
+      );
+    this.node = node;
+    this.ideas = ideas;
+    this.delimiter = delimiter;
+  }
+
+  /**
+   * Returns array filtered ideas, leaving only ideas that are not valid.
+   *  
+   * @param ideas - 
+   * @returns 
+   */
+  private listProblemParsedObjIdeas(ideas: IdeasItem[]) {
+    // todo: parse, don’t validate could be applied 
+    return ideas.filter(idea => {
+      if (idea instanceof ParsedObj) return false;
+      if (typeof idea === 'string' && /^\s+$/.test(idea)) return false;
+      if (Array.isArray(idea) && this.ideaItemsAreValid(idea)) return false;
+      return true;
+    });
+  }
+
+  /**
+   * Determine whether array only
+   *  
+   * @param items - Array of strings and Elements
+   * @returns Boolean
+   */
+  private ideaItemsAreValid(items: (ParsedObj|string|Element)[]) {
```

Comment:
I think this may be obsoleted by stronger typing (perhaps it was only useful at a point back where there was some bug in the produced objects), but maybe it may cover some instability in the dom output. Let’s remove it only after we have a more test covered codebase.

Reply:
Not much… I think that using the types from JSDOM should be enough to uncover problems and make everything easier to refactor and to simplify most of those mixed input functions. (So much for the possible switch to cheerio — but I find the DOM API easier to track anyway.)